app-misc/qlcplus: New proxy maintainer and version bump

kripton opened the pull request:
Adds myself as proxy maintainer
Add latest 4.x release (4.12.1)
Add latest 5.x alpha (5.0.0_alpha3)

kripton commented:
> Adding 5.0.0_alpha3 with keywords will make sure that almost no one is going to test 4.12.1. Is it worth it?

I'd like to have added 5.0.0_alpha3 with "~*" or "**" keyword to denote that it is a pre-release version. But since all Gentoo documentation pages I've found only allow stable or unstable keywords, I've decided to push both versions and have 4.12.1 moved to stable as soon as possible.
This way, everyone can try the version of QLC+ that he/she prefers.

a17r commented:
You could very well add it with empty KEYWORDS.

kripton commented:
Good idea, I'll do that :)

a17r commented on app-misc/qlcplus/qlcplus-4.12.1.ebuild:
```diff
@@ -0,0 +1,63 @@
+
+EAPI=6
+
+inherit qmake-utils udev vcs-snapshot virtualx xdg
```
This has a proper tarball, so the use of `vcs-snapshot` is dubious. If need be, set `S` manually.

kripton replied:
done

a17r commented on app-misc/qlcplus/qlcplus-4.12.1.ebuild:
```diff
@@ -0,0 +1,63 @@
+	dev-qt/qtmultimedia:5[widgets]
+	dev-qt/qtnetwork:5
+	dev-qt/qtscript:5
+	dev-qt/qttest:5
```
are you sure this is a runtime-depend?

kripton replied:
no binary installed links against it -> moved to DEPEND

a17r commented on app-misc/qlcplus/qlcplus-4.12.1.ebuild:
```diff
@@ -0,0 +1,63 @@
+
+	# sandbox error "mkdir /usr/share/qlcplus"
+	# see https://bugs.gentoo.org/621500#c2
+	export INSTALL_ROOT="${D}"
```
is this still necessary? under normal circumestances I would expect this as `emake` parameter in `src_install`.

kripton replied:
done. made it a parameter to emake

a17r commented on app-misc/qlcplus/qlcplus-4.12.1.ebuild:
```diff
@@ -0,0 +1,63 @@
+# Copyright 1999-2019 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=6
```
(also, please bump this to `EAPI=7`)

kripton replied:
done

a17r commented on app-misc/qlcplus/qlcplus-5.0.0_alpha3.ebuild:
```diff
@@ -0,0 +1,72 @@
+}
+
+pkg_postinst() {
+	elog "Some configurations of KDE break the layout of QLC+ 5's QML"
```
There are no versions of `KDE`, it is a community. Maybe `KDE Plasma` is really meant there?

kripton replied:
done

kripton commented:
@a17r : Done :) All nice and tidy in 3 commits

kripton commented:
Rebased again, this time with commit signing activated. Sorry, the signature got lost on the previous rebase.

mgorny commented on app-misc/qlcplus/qlcplus-4.12.1.ebuild:
```diff
@@ -0,0 +1,64 @@
+IUSE=""
+
+RDEPEND="
+	|| ( dev-embedded/libftdi:1 dev-embedded/libftdi:0 )
```
How is this library used? If the program links to it, then `||` is definitely wrong as you can't swap them without rebuilding it. Also, what happens if you have both slots installed?

kripton replied:
The build system automatically detects the version of libftdi installed. In case only SLOT 0 is installed, it is used. If libftdi:1 or both SLOTS are installed, libftdi-1.* is used.
You are right that they cannot be swapped without rebuilding but the software works when linked against either of them.
Could you propose a better solution to handle this in the ebuild? Thanks

mgorny replied:
`dev-embedded/libftdi:=`. This will select the newest installed slot, and force rebuild when newer slot is installed.

kripton replied:
Done. However, this approach might lead to problems when a new SLOT of dev-embedded/libftdi is introduced and downstream packages are not yet compatible with it.
For the moment, it should be fine :)

mgorny commented on app-misc/qlcplus/qlcplus-4.12.1.ebuild:
```diff
@@ -0,0 +1,64 @@
+
+src_prepare() {
+	default
+	sed -e "s:/etc/udev/rules.d:${EROOT}lib/udev/rules.d:" -i \
```
In EAPI 7, EROOT no longer carries implicit slash. Also I don't think EROOT is valid here. What is the path used for?

kripton replied:
It's a leftover from an old solution to install the udev rules to the correct location.
Since this is now handled via the following lines modifying the contents of variables.pri, this sed command is no longer used and I'll remove it in the next commits

kripton replied:
Done

mgorny commented on app-misc/qlcplus/qlcplus-4.12.1.ebuild:
```diff
@@ -0,0 +1,64 @@
+}
+
+src_install() {
+	emake DESTDIR="${D}" INSTALL_ROOT="${D}" install
```
Are both DESTDIR and INSTALL_ROOT used?

kripton replied:
I've tried all permutations (neither defined, only DESTDIR defined, only INSTALL_ROOT defined or both defined).
Both defined or only INSTALL_ROOT defined have the same result. If INSTALL_ROOT is not given (no matter if DESTDIR is used), a sandbox violation is reported.
I'll remove the DESTDIR parameter in the next commits. (Waiting for your answer regarding the libftdi dependency)

kripton replied:
Done

mgorny commented:
Thanks. I think it looks good. However, the tests failed for me:

```
********* Start testing of RGBScript_Test *********
Config: Using QtTest library 5.12.4, Qt 5.12.4 (x86_64-little_endian-lp64 shared (dynamic) release build; by GCC 8.3.0)
PASS   : RGBScript_Test::initTestCase()
PASS   : RGBScript_Test::initial()
PASS   : RGBScript_Test::directories()
QFATAL : RGBScript_Test::scripts() Cannot mix incompatible Qt library (version 0x50b03) with this library (version 0x50c04)
FAIL!  : RGBScript_Test::scripts() Received a fatal error.
   Loc: [Unknown file(0)]
Totals: 3 passed, 1 failed, 0 skipped, 0 blacklisted, 84ms
********* Finished testing of RGBScript_Test *********
./test.sh: line 4:   519 Aborted                 (core dumped) ./rgbscript_test
```

Full log:  [app-misc:qlcplus-4.12.1:20190729-083025.log](https://github.com/gentoo/gentoo/files/3441181/app-misc.qlcplus-4.12.1.20190729-083025.log)

I suppose it's some outdated dep on my system. I'll try to find and upgrade it.

I don't know if it's something that can be solved in ebuild at the moment or not. @gentoo/qt ?

mgorny commented:
I'm afraid that after resolving this, one test still fails, and test suite crashes a while later:
[app-misc:qlcplus-4.12.1:20190729-093402.log](https://github.com/gentoo/gentoo/files/3441328/app-misc.qlcplus-4.12.1.20190729-093402.log)

Pesa commented:
@mgorny what dep was it?

mgorny commented:
> @mgorny what dep was it?

qtscript was older (5.11*) than other qt packages (5.12*).

kripton commented:
@mgorny: Yes, Qt is a bit picky when mixing version ;)

About the failed test: Yes, I can confirm that on my machine. However, it seems to be related to compiler version, judging from this commit from upstream: https://github.com/mcallegari/qlcplus/commit/fe2b812576ed69f1d2aab5a0f940d09db9400b7c

So it looks like it should be fixed in the next release.

Unfortunately, I cannot run the tests with qlcplus-4.11.1 (latest version in gentoo tree) since that version doesn't even compile on my system (tested with gcc 8.3.0 and gcc 9.1.0). So, even assuming that the failed test is not a regression, I'd see this PR as an improvement of the current situation (latest version in tree not compiling at all).

I'd propose to either:
* wait for the next release from upstream. I'd update this PR in that case
* merge this PR and I'll make a new one (and run make test before submitting it) whan upstream fixed that issue

In addition, I could also provide live ebuilds for -4.9999 and -5.9999 so we can check the failing tests right now

a17r commented:
ping @mgorny

mgorny commented:
@kripton, are you sure that the newer version actually fixes it? Note that the test suite also crashes for me which is weird, and it's unclear if it's related or not. If you really want to commit it as-is, please look into skipping/xfailing that test.

kripton commented:
@mgorny: The latest git HEAD for 4.x builds and tests fine:
```
/var/tmp/portage/app-misc/qlcplus-4.9999/work/qlcplus-4.9999
Unit tests passed.
>>> Completed testing app-misc/qlcplus-4.9999```

As I said, I can upload the live ebuilds as well, but I'm unsure if this PR is better without them and the live ebuilds should be in a separate one?

I will check on how to disable the specific test cases via sed or similar. Alternatively, I could extract that specific fix and supply it as patch via the files dir. Which approach would you prefer?

kripton commented:
Upstream released qlcplus-4.12.2 yesterday. I've updated the ebuild and the tests are no longer failing or crashing.
I've also discovered that dev-qt/linguist-tools:5 is required to build the app. I've added this as BDEPEND to the ebuild for 4.12.1 and 5.0.0_alpha3.

@mgorny : Looking forward to your feedback and/or merge.

kripton commented:
ping @mgorny

juippis commented:
(a green light from the CI bot is required, so you might have to rebase against current fresh gentoo master)

kripton commented:
@juippis: Thanks for the note. I saw the CI errors but thought since it is obvious that it is caused by other ebuilds, no one would care. Rebased, waiting for CI now :)

mgorny commented on app-misc/qlcplus/qlcplus-4.12.2.ebuild:
```diff
@@ -0,0 +1,63 @@
+		-i variables.pri || die
+
+	## Remove Werror-flag since there are some warnings with gcc-9.x
+	sed -e "s/QMAKE_CXXFLAGS += -Werror/#QMAKE_CXXFLAGS += -Werror/g" \
```
Either `/QMAKE_CXXFLAGS += -Werror/d` or `s/.../#&/` is shorter (`&` meaning match from LHS).

kripton replied:
I'll investigate, thanks for the hint

kripton replied:
Changed to use `s/.../#&/g`

mgorny commented on app-misc/qlcplus/qlcplus-4.12.2.ebuild:
```diff
@@ -0,0 +1,63 @@
+
+EAPI=7
+
+inherit qmake-utils udev virtualx xdg
```
How is `udev` used?

kripton replied:
It's used to query the path where to put the udev rules here: https://github.com/gentoo/gentoo/pull/12418/files#diff-6045b184eacc87c5adb4beb2ec13ffe3R45

mgorny commented on app-misc/qlcplus/qlcplus-4.12.2.ebuild:
```diff
@@ -0,0 +1,63 @@
+
+DESCRIPTION="A software to control DMX or analog lighting systems"
+HOMEPAGE="https://www.qlcplus.org/"
+SRC_URI="https://github.com/mcallegari/${PN}/archive/QLC+_${PV}.tar.gz -> ${P}.tar.gz"
```
Why are you renaming it? The original name looks fine.

kripton replied:
That was copied from the ebuild currently in the portage tree, I just didn't bother to change. Of course we can leave away the renaming if it doesn't break anything. Will investigate and change

kripton replied:
Changed ebuilds to not rename the distfile and confirmed that it works properly

mgorny commented on app-misc/qlcplus/qlcplus-4.12.2.ebuild:
```diff
@@ -0,0 +1,63 @@
+	media-libs/libmad
+	media-libs/libsndfile
+	sci-libs/fftw:3.0
+	virtual/libusb:0
```
Does it really simultaneously use both version of libusb? oO

kripton replied:
Yep, I just confirmed that. QLC+ ships with several plugins for DMX input/output devices. Some plugins use the current libusb (the plugin for FTDI-based devices), some use the legacy one (the one for "Peperoni"-based devices or uDMX-devices). The plugins might have been written at different times and by different authors and upstream didn't harmonize them yet

kripton commented:
Ping @mgorny

kripton commented:
Ping @mgorny
Anything that's blocking review and I could to to help? Thanks!

juippis commented:
```
x86_64-pc-linux-gnu-g++ -c -march=native -O2 -pipe -frecord-gcc-switches -Wno-unused-local-typedefs -Wall -W -D_REENTRANT -fPIC -DQT_NO_DEBUG_OUTPUT -DQT_NO_DEBUG -DQT_GUI_LIB -DQT_TESTLIB_LIB -DQT_CORE_LIB -DQT_TESTCASE_BUILDDIR='"/tmp/portage/app-misc/qlcplus-4.12.2/work/qlcplus-QLC-_4.12.2/engine/test/qlcchannel"' -I. -I../../../plugins/interfaces -I../../src -isystem /usr/include/qt5 -isystem /usr/include/qt5/QtGui -isystem /usr/include/qt5/QtTest -isystem /usr/include/qt5/QtCore -I. -isystem /usr/include/libdrm -I/usr/lib64/qt5/mkspecs/linux-g++ -o moc_qlcchannel_test.o moc_qlcchannel_test.cpp
/usr/lib/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: chaser_test.o: in function `main':
chaser_test.cpp:(.text.startup+0x54): undefined reference to `qgpu_features(QString const&)'
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:150: chaser_test] Error 1
make[3]: Leaving directory '/tmp/portage/app-misc/qlcplus-4.12.2/work/qlcplus-QLC-_4.12.2/engine/test/chaser'
make[2]: *** [Makefile:114: sub-chaser-make_first] Error 2
make[2]: *** Waiting for unfinished jobs....

/usr/lib/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: qlci18n_test.o: in function `main':
qlci18n_test.cpp:(.text.startup+0x54): undefined reference to `qgpu_features(QString const&)'
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:143: qlci18n_test] Error 1
make[3]: Leaving directory '/tmp/portage/app-misc/qlcplus-4.12.2/work/qlcplus-QLC-_4.12.2/engine/test/qlci18n'
make[2]: *** [Makefile:764: sub-qlci18n-make_first] Error 2

make[3]: Leaving directory '/tmp/portage/app-misc/qlcplus-4.12.2/work/qlcplus-QLC-_4.12.2/engine/test/qlcinputchannel'
/usr/lib/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: mastertimer_test.o: in function `main':
mastertimer_test.cpp:(.text.startup+0x54): undefined reference to `qgpu_features(QString const&)'
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:157: mastertimer_test] Error 1
make[3]: Leaving directory '/tmp/portage/app-misc/qlcplus-4.12.2/work/qlcplus-QLC-_4.12.2/engine/test/mastertimer'
make[2]: *** [Makefile:539: sub-mastertimer-make_first] Error 2
make[3]: Leaving directory '/tmp/portage/app-misc/qlcplus-4.12.2/work/qlcplus-QLC-_4.12.2/engine/test/qlcfixturemode'
make[2]: Leaving directory '/tmp/portage/app-misc/qlcplus-4.12.2/work/qlcplus-QLC-_4.12.2/engine/test'
make[1]: *** [Makefile:135: sub-test-make_first-ordered] Error 2
make[1]: Leaving directory '/tmp/portage/app-misc/qlcplus-4.12.2/work/qlcplus-QLC-_4.12.2/engine'
make: *** [Makefile:98: sub-engine-make_first-ordered] Error 2
 * ERROR: app-misc/qlcplus-4.12.2::testworld failed (compile phase):
 *   emake failed
```
I think some dep is missing?

kripton commented:
Hi @juippis , thanks for testing the ebuild.
I'm not sure if it's really a missing dependency since it fails at link time. So all headers for the compile step are there but somehow the linker fails nevertheless.
Could you please share your complete build.log somewhere (as a gist?) and I will investigate in parallel on my machines. Thanks!

juippis commented:
https://dev.gentoo.org/~juippis/logs/qlcplus-4.12.2_build.log

kripton commented:
@juippis   To be honest, I'm not able to reproduce. After first research, I thought that `qgpu_features(QString const&)` was related to QtTest (search for it on http://darnassus.sceen.net/~teythoon/gnupg/covererage/201703021637/gpgme/usr/include/x86_64-linux-gnu/qt5/QtTest/qtest.h.gcov.html for example).
So I forcefully unmerged qttest and tried to compile the ebuild => failed but with a different error.
Then, I thought it might be related to qt3d since the function name included `gpu` => qlcplus-4.12.2 compiles and installed fine.

Then, to be very sure and independent of my local environment, I ran this `Dockerfile`:
```
# name the portage image
FROM gentoo/portage:latest as portage

# image is based on stage3-amd64-nomultilib
FROM gentoo/stage3-amd64-nomultilib:latest

# copy the entire portage volume in
COPY --from=portage /var/db/repos/gentoo /var/db/repos/gentoo

# Prepare the qlcplus ebuild
RUN \
        cd /var/db/repos/gentoo && \
        mkdir -p app-misc/qlcplus && \
        cd app-misc/qlcplus && \
        wget https://github.com/kripton/gentoo/raw/qlcplus_proxymaint/app-misc/qlcplus/qlcplus-4.12.2.ebuild && \
        ebuild qlcplus-4.12.2.ebuild digest

# Reduce dependencies
RUN \
        echo USE=\"-* python_single_target_python3_6 python_targets_python3_6 unicode xcb\" >> /etc/portage/make.conf && \
        echo VIDEO_CARDS=\"\" >> /etc/portage/make.conf && \
        echo FEATURES=\"parallel-fetch collision-protect\" >> /etc/portage/make.conf && \
        cat /etc/portage/make.conf

# emerge it
RUN \
        echo =app-misc/qlcplus-4.12.2::gentoo > /etc/portage/package.keywords && \
        echo dev-libs/openssl bindist >> /etc/portage/package.use/qlcplus && \
        echo dev-qt/qtnetwork bindist >> /etc/portage/package.use/qlcplus && \
        echo dev-qt/qtmultimedia  widgets >> /etc/portage/package.use/qlcplus && \
        echo x11-libs/libxcb xkb >> /etc/portage/package.use/qlcplus && \
        echo x11-libs/libxkbcommon X >> /etc/portage/package.use/qlcplus && \
        echo dev-libs/libpcre2 pcre16 >> /etc/portage/package.use/qlcplus && \
        emerge -v --job=3 qlcplus
```
and qlcplus-4.12.2 emerged fine there.

This leads me to the conclusion that something in your Qt-environment might be broken. Maybe due to a GCC ABI breakage, maybe due to a partial Qt update? Anything else that might be special or uncommon that I didn't think about?

Would you mind re-emerging all your installed Qt-packages on your system and try again? Something such as  `emerge -1av $(eix -I -C dev-qt --only-names)` should do.

Of course you could also run that Dockerfile on your machine to check.

Thank you!

juippis commented:
Indeed that seemed to have caused it. Thanks for delving deeply into this. 

I usually check these packages on a fresh base system install (image), but now I've been using one with webkit-gtk and qtwebengine pre-installed. Need to update my base image and re-think my strategy...

kripton commented:
Sure, no problem. Thanks for re-testing and merging :D :+1:
